Convert ac_breakout to using python configurable

breakds opened the pull request:
# Description

This converts the original `.gin` file to the configuration file. The original motivation is to test the atari based gym.

# Integration Test

Trained on both the `.gin` and `_conf.py` and the comparison looks like below:

Left `_confg.py` vs Right `.gin`

![comparison_ac_breakout](https://user-images.githubusercontent.com/1111035/124833507-23177c80-df33-11eb-8d57-51c92759a1c5.png)

hnyu commented:
Probably run each conf file longer to confirm the performance is good enough? The "AverageReturn" metric should be above 250 after a while.

breakds committed:
Convert ac_breakout to using python configurable

breakds commented:
> Probably run each conf file longer to confirm the performance is good enough? The "AverageReturn" metric should be above 250 after a while.

Acknowledged. Will run training for longer time and monitor.

breakds commented:
After some more extended training, evaluating and digging I think I am close to what have happened. The `AverageReturn` I saw from yesterday was misleading because in the tensorboard graph the range of the `x`-axis (denoting the number of episode steps is only 1k), and that is why it reaches only around `2.5`.

I have no idea how I can tweak the visualization to show more than `1k` episode steps during training. However, I found that if I call `alf.bin.play` for an extensive time I can observe:

1. First, the trained agent is actually pretty good, and it reaches score = 400 easily on every game (and then get stuck in the dead loop).
2. After the `play`, if I look at the tensorboard again, it correctly shows `AverageReturn` with 50k steps or so. And the Average Return looks like below:

![average_return](https://user-images.githubusercontent.com/1111035/125002422-b91dd680-e009-11eb-9ad0-2c06c7258730.png)

I think this also concludes the integration test of this PR.

hnyu commented:
> 2. After the `play`, if I look at the tensorboard again, it correctly shows `AverageReturn` with 50k steps or so. And the Average Return looks like below:

The play shouldn't affect how tensorboard shows. Can you share me with your tensorboard link?

breakds commented:
> > 1. After the `play`, if I look at the tensorboard again, it correctly shows `AverageReturn` with 50k steps or so. And the Average Return looks like below:
> 
> The play shouldn't affect how tensorboard shows. Can you share me with your tensorboard link?

Yep sure, are you coming to the office tomorrow? I can demonstrate that and consult you on this topic.

hnyu commented:
> > > 1. After the `play`, if I look at the tensorboard again, it correctly shows `AverageReturn` with 50k steps or so. And the Average Return looks like below:
> > 
> > 
> > The play shouldn't affect how tensorboard shows. Can you share me with your tensorboard link?
> 
> Yep sure, are you coming to the office tomorrow? I can demonstrate that and consult you on this topic.

Sounds good

hnyu commented on alf/examples/ac_breakout_conf.py:
```diff
@@ -0,0 +1,70 @@
+# Neural Network Configuration
+CONV_LAYER_PARAMS = ((32, 8, 4), (64, 4, 2), (64, 3, 1))
+alf.config(
+    'ActorDistributionNetwork',
```
A better way of configuring networks would be using partial, to avoid configuring the network class in a global scope. 

```python
actor_network_cls = partial(ActorDistributionNetwork, fc_layer_params=(512,), conv_layer_params=CONV_LAYER_PARAMS)

alf.config(
    'ActorCriticAlgorithm',
    actor_network_ctor=actor_network_cls,
    ...)
``` 

Same for the value network.

hnyu commented on alf/examples/atari_conf.py:
```diff
@@ -0,0 +1,34 @@
+from alf.environments import suite_gym, gym_wrappers
+from alf.algorithms.data_transformer import FrameStacker, ImageScaleTransformer, RewardClipping
+
+alf.config('create_environment', env_load_fn=suite_gym.load)
```
Would be better specifying a max episode length:

```python
alf.config('suite_gym.load', max_episode_steps=4500) # DQN paper setting, 18000 frames assuming frameskip=4
```

breakds commented:
Updated the PR addressing @hnyu 's comments:

1. Removing the original `.gin` file
2. Using `functools.partial` for both the actor network and value network
3. Default max episode steps to 4500 (originally implicitly set as 400,000)
4. Also, updated `README.md` to link to the python conf version

Will run training again and update the thread.

breakds commented:
Ran the experiment with the updated commit and the result is similar. Was able to finish training 5M steps in 22 minutes.

hnyu commented on alf/bin/train_play_test.py:
```diff
@@ -319,6 +320,7 @@ def _test_train(self, conf_file, extra_params, root_dir):
         if conf_file.endswith('.gin'):
             cmd.append('--gin_file=%s' % conf_file)
         else:
+            print(f'conf_file={conf_file}')
```
remove this print

breakds replied:
Thought I have removed this. Sorry!

hnyu reviewed (changes requested):
Also when you make additional changes to a PR, please do not rebase and force pushing. Just push on top of the existing commits so that the delta change can be seen on github.

breakds committed:
Remove unnecessary print

breakds commented:
@hnyu Appreciate the thorough review!
